internal/gengapic/genrest.go

```diff
@@ -210,33 +210,99 @@ func (g *generator) queryParams(m *descriptor.MethodDescriptorProto) map[string]
-	}
 
-	for _, f := range request.GetField() {
-		// If, and only if, a field is not a path parameter or a body parameter,
+	// Possible query parameters are all leaf fields in the request or body.
+	pathToLeaf := g.getLeafs(request, bodyField)
+	// Iterate in sorted order to
+	for path, leaf := range pathToLeaf {
+		// If, and only if, a leaf field is not a path parameter or a body parameter,
 		// it is a query parameter.
-		if _, ok := pathParams[f.GetName()]; !ok && g.lookupField(bodyMsg.GetName(), f.GetName()) == nil {
-			queryParams[f.GetName()] = f
+		if _, ok := pathParams[path]; !ok && g.lookupField(request.GetName(), leaf.GetName()) == nil {
+			queryParams[path] = leaf
 		}
 	}
 
 	return queryParams
 }
 
+// Returns a map from fully qualified path to field descriptor for all the leaf fields of a message 'm',
+// where a "leaf" field is a non-message whose top message ancestor is 'm'.
+// e.g. for a message like the following
+//
+// message Mollusc {
+//     message Squid {
+//         message Mantle {
+//             int32 mass_kg = 1;
+//         }
+//         Mantle mantle = 1;
+//     }
+//     Squid squid = 1;
+// }
+//
+// The one entry would be
+// "squid.mantle.mass_kg": *descriptor.FieldDescriptorProto...
+func (g *generator) getLeafs(msg *descriptor.DescriptorProto, excludedFields ...*descriptor.FieldDescriptorProto) map[string]*descriptor.FieldDescriptorProto {
+	pathsToLeafs := map[string]*descriptor.FieldDescriptorProto{}
+
+	isInSliceP := func(fields []*descriptor.FieldDescriptorProto, field *descriptor.FieldDescriptorProto) bool {
+		for _, f := range fields {
+			if field == f {
+				return true
+			}
+		}
+		return false
+	}
+
+	// We need to declare and define this function in two steps
+	// so that we can use it recursively.
+	var recurse func([]*descriptor.FieldDescriptorProto, *descriptor.DescriptorProto)
+
+	recurse = func(
+		stack []*descriptor.FieldDescriptorProto,
+		m *descriptor.DescriptorProto,
+	) {
+		for _, field := range m.GetField() {
+			if field.GetType() == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
+				if field.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
+					// Repeated message fields must not be mapped because no
+					// client library can support such complicated mappings.
+					// https://cloud.google.com/endpoints/docs/grpc-service-config/reference/rpc/google.api#grpc-transcoding
+					continue
+				}
+				if isInSliceP(excludedFields, field) {
+					continue
+				}
+				// Short circuit on infinite recursion
+				if isInSliceP(stack, field) {
+					continue
+				}
+
+				subMsg := g.descInfo.Type[field.GetTypeName()].(*descriptor.DescriptorProto)
+				recurse(append(stack, field), subMsg)
+			} else {
+				elts := []string{}
+				for _, f := range stack {
+					elts = append(elts, f.GetName())
+				}
+				elts = append(elts, field.GetName())
+				key := strings.Join(elts, ".")
+				pathsToLeafs[key] = field
+			}
```

Comment:
This is total nit, feel free to ignore: I (as does Go style) prefer to reduce indentation where possible. `if-else` chains are a good opportunity for that. As a matter of preference, I'd suggest inverting the `if-else` to be `if!` + fall through. Personally I think its more readable but 🤷🏻‍♂️ 

```go
if field.GetType() != descriptor.FieldDescriptorProto_TYPE_MESSAGE {
  elts := []string{}
  // ...

  // this continue is critical to avoid the fall through into the Message type logic.
  continue
}

if field.GetLabel() == REPEATED { continue }

if contains(excludedFields, field) { continue }

if contains(stack, field) { continue }

subMsg := ...
recurse...
```

Reply:
In this one particular case I think I'd like to keep the base case and the recursive case more firmly distinct. IMO it makes for a better match to a recursive traversal pattern. I'll try rearranging with less indentation to see how it looks, or maybe refactor the two cases into their own standalone functions.

Reply:
Eh leave it the way it is in that case. I think that's a fair reason. Any more refactoring and it might end up getting too complicated. :)

Reply:
Okay, I don't know how idiomatic the new version is, but it definitely reduces indentation a little bit.